### DISCUSSION TITLE
[jsonwebtoken] Why does `verify` return a string?

### DISCUSSION THREAD

**Discussion by paulius-valiunas** [DefinitelyTyped/DefinitelyTyped#56954] [Issues with a @types package] [answered]
From latest `@types/jsonwebtoken`:

```ts
export function verify(token: string, secretOrPublicKey: Secret, options: VerifyOptions & { complete: true }): Jwt | string;
export function verify(token: string, secretOrPublicKey: Secret, options?: VerifyOptions): JwtPayload | string;
```

Is there a specific case where `verify` is supposed to return a string rather than `Jwt` or `JwtPayload`? I don't see any mention of that in the documentation and I also don't see how it could possibly return a string when looking at the source code:
https://github.com/auth0/node-jsonwebtoken/blob/master/verify.js

[a comment by an automated account is not shown]

**Reply by peterblazejewicz** [reply]
it was introduced in #53435 I think there was a mistake done in PR, as its changes were based on how JTW token validator should work (via RFC ref and rewritten part of @types/jsonwebtoken) without referencing actual JavaScript package implementation (which can be different and not 100% aligned with RFC with details).
Please fill free to correct this detail via PR

**Reply by nandi95** [reply]
The string [was there](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/53435/files#diff-37161fa05c1ae7e10456ed1c324e607f937ea35be4d2c920c0c289a67e94f131L180) beforehand. I don't think [verify](https://github.com/auth0/node-jsonwebtoken/blob/master/verify.js) ever returns a string nor does the [docs](https://github.com/auth0/node-jsonwebtoken#jwtverifytoken-secretorpublickey-options-callback) indicates so.

**Reply by peterblazejewicz** [reply]
yes, sorry, I should use proper wording: we missed opportunity to catch the problem when doing #53435 review (@nandi95 sry!)

**Reply by nwjsmith** [reply]
@peterblazejewicz opened a PR here https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58126

**Comment by peterblazejewicz** [comment, marked as answer]
Fixed in  #58126

**Comment by LinusU** [comment]
The payload will be a string for all tokens that have a payload that isn't JSON. It's a bit hard to follow, but if you look closely at the verify source code you can see that can indeed return a string:

https://github.com/auth0/node-jsonwebtoken/blob/74d5719bd03993fcf71e3b176621f133eb6138c0/verify.js#L66-L69

https://github.com/auth0/node-jsonwebtoken/blob/74d5719bd03993fcf71e3b176621f133eb6138c0/decode.js#L9-L17

https://github.com/auth0/node-jsonwebtoken/blob/74d5719bd03993fcf71e3b176621f133eb6138c0/verify.js#L136

https://github.com/auth0/node-jsonwebtoken/blob/74d5719bd03993fcf71e3b176621f133eb6138c0/verify.js#L213-L223

It should however never return a string when `completed: true` is passed, in that case the `payload` property of the returned object should be a string. I've submitted a PR that addresses all of this here: #58158
